### PR TITLE
Return an error when a column does not exist in window function

### LINKED ISSUE #9166: Invalid column reference in window function order by causes panic
### Describe the bug

[This unwrap](https://github.com/apache/arrow-datafusion/blob/main/datafusion/sql/src/expr/function.rs#L156) causes a panic when the referenced column doesn't exist in the schema.

### To Reproduce

In datafusion-cli v35.0.0

```DataFusion CLI v35.0.0
❯ select count() over (order by a);
thread 'main' panicked at datafusion-sql-35.0.0/src/expr/function.rs:112:63:
called `Result::unwrap()` on an `Err` value: SchemaError(FieldNotFound { field: Column { relation: None, name: "a" }, valid_fields: [] }, Some(""))
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

### Expected behavior

A graceful error instead of a panic.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by PhVHoang**
## Which issue does this PR close?
Closes #9166.

## Rationale for this change
To return an error instead of panicking when columns don't exist in window function.

## What changes are included in this PR?
Code, tests

## Are these changes tested?
Yes

### CODE REVIEW

File: datafusion/sql/src/expr/function.rs [hunk 1]
```diff
@@ -149,18 +149,37 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
+            // Find whether ties are possible in the given ordering
+            let is_ordering_strict = order_by.iter().find_map(|orderby_expr| {
                 if let Expr::Sort(sort_expr) = orderby_expr {
                     if let Expr::Column(col) = sort_expr.expr.as_ref() {
-                        let idx = schema.index_of_column(col).unwrap();
-                        return func_deps.iter().any(|dep| {
+                        let idx_result = schema.index_of_column(col);
+
+                        // Propagate the error to the caller and return it immediately
+                        let idx = idx_result
```
Comment (alamb) [inline datafusion/sql/src/expr/function.rs:159]: I think the `ok()` function call here discards the error. 

I think changing 
```rust
                        let idx = schema.index_of_column(col).unwrap();
```

To 
```rust
                        let idx = schema.index_of_column(col).ok()?;
```

accomplishes the same thing (and your test still passes)

**Review by alamb — COMMENTED** [review]
Thank you @PhVHoang  -- this is a great first contribution. 

I think the code can be made simpler and I left a comment for your consideration

File: datafusion/sql/src/expr/function.rs (hunk 1 above)
Comment (PhVHoang) [reply datafusion/sql/src/expr/function.rs]: Thank you very much for your suggestion. Yes it's way simpler than that of mine. 

Fixed.

**Review by comphead — APPROVED** [review]
lgtm thanks @PhVHoang 
its better than crash

**Comment by comphead** [comment]
@PhVHoang please add the test also for non existent partition by column
and qualified column like `t.a`, `schema.t.a`, `catalog.schema.t.a`

**Review by mustafasrepo — APPROVED** [review]
Thanks @PhVHoang for this PR. LGTM!

**Comment by alamb** [comment]
> @PhVHoang please add the test also for non existent partition by column and qualified column like `t.a`, `schema.t.a`, `catalog.schema.t.a`

I think the only thing remaining before merging this PR is the suggestion from @comphead https://github.com/apache/arrow-datafusion/pull/9202#issuecomment-1939321659 for a few more tests. I can add them hopefully later today if no one else has a chance

**Comment by PhVHoang** [comment]
@alamb @comphead I added some more tests. Please take a look and let me know if these tests are sufficient for this change.

**Review by comphead — APPROVED** [review]
lgtm thanks @PhVHoang

**Pull request merged**